bot/cogs/filter_lists.py

```diff
@@ -0,0 +1,250 @@
+        else:
+            await ctx.message.add_reaction("❌")
+
+    async def _list_all_data(self, ctx: Context, allowed: bool, list_type: ValidFilterListType) -> None:
+        """Paginate and display all items in a filterlist."""
+        allow_type = "whitelist" if allowed else "blacklist"
+        result = self.bot.filter_list_cache[f"{list_type}.{allowed}"]
+
+        # Build a list of lines we want to show in the paginator
+        lines = []
+        for content, metadata in result.items():
+            line = f"• `{content}`"
+
+            if metadata.get("comment"):
+                line += f" - {metadata.get('comment')}"
+
+            lines.append(line)
+        lines = sorted(lines)
+
+        # Build the embed
+        list_type_plural = list_type.lower().replace("_", " ").title() + "s"
+        embed = Embed(
+            title=f"{allow_type.title()}ed {list_type_plural} ({len(result)} total)",
+            colour=Colour.blue()
+        )
+        log.trace(f"Trying to list {len(result)} items from the {list_type.lower()} {allow_type}")
+
+        if result:
+            await LinePaginator.paginate(lines, ctx, embed, max_lines=15, empty=False)
+        else:
+            embed.description = "Hmmm, seems like there's nothing here yet."
+            await ctx.send(embed=embed)
+            await ctx.message.add_reaction("❌")
+
+    @staticmethod
+    async def _validate_guild_invite(ctx: Context, invite: str) -> dict:
+        """
+        Validates a guild invite, and returns the guild info as a dict.
+
+        Will raise a BadArgument if the guild invite is invalid.
+        """
+        log.trace(f"Attempting to validate whether or not {invite} is a guild invite.")
+        validator = ValidDiscordServerInvite()
+        guild_data = await validator.convert(ctx, invite)
+
+        # If we make it this far without raising a BadArgument, the invite is
+        # valid. Let's return a dict of guild information.
+        log.trace(f"{invite} validated as server invite. Converting to ID.")
+        return guild_data
+
+    @group(aliases=("allowlist", "allow", "al", "wl"))
+    async def whitelist(self, ctx: Context) -> None:
+        """Group for whitelisting commands."""
+        if not ctx.invoked_subcommand:
+            await ctx.send_help(ctx.command)
+
+    @group(aliases=("denylist", "deny", "bl", "dl"))
+    async def blacklist(self, ctx: Context) -> None:
+        """Group for blacklisting commands."""
+        if not ctx.invoked_subcommand:
+            await ctx.send_help(ctx.command)
+
+    @whitelist.command(name="add", aliases=("a", "set"))
+    async def allow_add(
+            self,
+            ctx: Context,
+            list_type: ValidFilterListType,
+            content: str,
+            *,
+            comment: Optional[str] = None,
+    ) -> None:
+        """Add an item to the specified allowlist."""
+        await self._add_data(ctx, True, list_type, content, comment)
```

Comment:
I think prevailing code-style is to indent at 4 spaces always.

Reply:
On one hand, that is true (though there is also some existing code with 8 spaces) and my personal preference. On the other hand, both PEP 8 and our own style guide suggest using 8 spaces.

Reply:
I can change it - this was probably changed by my IDE, not a conscious choice - but I do think both approaches have clear advantages, and that there's no clear winner. This style makes a clearer visual separation between the paramters and the function body, and yeah, PEP8 does suggest it. On the other hand, PEP8 also says "adhere to prevailing style", so no clear winner.

Reply:
ugh. I changed it and I really think it makes it worse. but oh well.